routers/web/repo/issue.go

```diff
@@ -1020,32 +1020,40 @@ func commentTag(repo *models.Repository, poster *models.User, issue *models.Issu
 	if err != nil {
 		return models.CommentTagNone, err
 	}
-	if perm.IsOwner() {
-		if !poster.IsAdmin {
-			return models.CommentTagOwner, nil
-		}
 
-		ok, err := models.IsUserRealRepoAdmin(repo, poster)
-		if err != nil {
-			return models.CommentTagNone, err
-		}
+	// By default the poster has no tags.
+	pCommentTag := models.CommentTagNone
 
-		if ok {
-			return models.CommentTagOwner, nil
-		}
+	// Check if the poster is owner of the repo.
+	if perm.IsOwner() {
+		// If the poster isn't a admin, enable the owner Tag.
+		if !poster.IsAdmin {
+			pCommentTag = pCommentTag.WithTag(models.CommentTagOwner)
+		} else {
 
-		if ok, err = repo.IsCollaborator(poster.ID); ok && err == nil {
-			return models.CommentTagWriter, nil
+			// Otherwise check if poster is the real repo admin.
+			ok, err := models.IsUserRealRepoAdmin(repo, poster)
+			if err != nil {
+				return models.CommentTagNone, err
+			}
+			if ok {
+				pCommentTag = pCommentTag.WithTag(models.CommentTagOwner)
+			}
 		}
+	}
 
-		return models.CommentTagNone, err
+	// Is the poster can write code to the repo, enable Writer tag.
+	// Only enable this if the poster doesn't have the owner tag already.
+	if !pCommentTag.HasTag("Owner") && perm.CanWrite(models.UnitTypeCode) {
```

Comment:
I think we should check `UnitTypeIssue` or `UnitTypePullRequest` but not `UnitTypeCode`. Assume one person is an issue manager but don't have permissions to `Code`.

Reply:
Otherwise LGTM.

Reply:
I think this should be moved to another Issue or PR as this is a simple refactor of the code. I don't think it's within this scope of the PR to edit such thing.

Reply:
OK. So please wait for #17591 merged.